### ISSUE TITLE
Support Math Notation in Articles

### ISSUE THREAD

**Issue by nestedsoftware** [thepracticaldev/dev.to#182]
# Feature
### Support Math Notation in Articles

It would be nice to be able to use math notation in articles rather than having to create a new image for each piece of mathematical notation. Perhaps something like [mathjax](https://github.com/mathjax/MathJax).

**Comment by Zhao-Andy** [comment]
Hey @nestedsoftware, thanks for the feedback. Going to backlog this for now. We had a similar request raised before: #97 Latex Support in Articles. If you have any experience with Latex, feel free to leave your thoughts in that issue, too.

**Comment by maestromac** [comment]
Reopening since this is definitely a useful feature.

**Comment by jessleenyc** [comment]
@benhalpern I don't remember where we left off with this. Do you recall?

**Comment by Zhao-Andy** [comment]
Not stale, still think this is pretty useful.

**Comment by wangonya** [comment]
I've been thinking of doing a series on "Solving math problems with python" and I think this would be very helpful. Might look into it and open a PR if possible.

**Comment by turbo** [comment]
I'd suggest using KaTeX (https://katex.org/) for supporting math notation. As an example of a Markdown editor implementing LaTex support, https://stackedit.io/ (https://github.com/benweet/stackedit) uses the following syntax (in Markdown):

```markdown
## KaTeX

You can render LaTeX mathematical expressions using [KaTeX](https://khan.github.io/KaTeX/):

The *Gamma function* satisfying $\Gamma(n) = (n-1)!\quad\forall n\in\mathbb N$ is via the Euler integral

$$
\Gamma(z) = \int_0^\infty t^{z-1}e^{-t}dt\,.
$$

> You can find more information about **LaTeX** mathematical expressions [here](http://meta.math.stackexchange.com/questions/5020/mathjax-basic-tutorial-and-quick-reference).
```

**Comment by turbo** [comment]
I too am using LaTeX equations in some of my blog posts I'd like to port over to dev.to, so this is blocking my worklfow FWIW. Though I see that this issue is marked with a low priority. What would be the best way to attack this (adding support for a new feature in Markdown) for an external contributor (like me, though not making any promises). Where (source files/directories) do you see modifications to be neccessary (KaTeX requires including at least the JS in the frontend)?

**Comment by rhymes** [comment]
hi @turbo, thanks for reviving this ticket. I don't believe we can easily use KaTex being it in Node. The parsing, and rendering and cleaning is all done server side (so no client side plugins) and I suspect it might be complicated to introduce a Node external process just to render math notation.

The engine we use is called [Redcarpet](https://rubygems.org/gems/redcarpet) and it doesn't support math notations and probably won't in the foreseeable future (there's a long discussion culminating with a rejection from 2014: https://github.com/vmg/redcarpet/issues/313#issuecomment-50665365).

There is a Ruby gem which could be used in place of Redcarpet - [kramdown](https://rubygems.org/gems/kramdown) - which seems to support everything we need (syntax highlighting for example) in addition MathJax natively (and other notations with plugins) but for the few minutes I spent looking at the matter (so I'm definitely not an expert on this) it seems there are a few items for discussion, see also the discussion about adding math notation support to Github itself - https://github.com/github/markup/issues/897:

- what notation is the most common? because I guess if we add support to one we have to think about supporting the others
- are there any security concerns with these notations?

**Comment by turbo** [comment]
> being it in Node

That's certainly not a requirement, and a bit misleading. KaTeX is simply JavaScript. It also doesn't involve any parsing of markup (server-side). You'd serve the KaTeX JS file, and by simply having LaTeX notation on the page, in the HTML / Text, it would be displayed as LaTeX. There is no extra step required to enable or parse KaTeX server-side. You can see this e.g. by opening StackEdit and copying the demo Markdown into GitHub or dev.to. The two missing features (Mermaid and LaTeX) are simply rendered verbatim, as code blocks and text respectively.

> no client side plugins

No server-side rendering would ever be able to correctly typeset LaTeX for all devices, resolutions and supported browser features (SVG rendering support levels, etc.). 

Plus, other shortcodes readily insert JS into dev.to already (like Gists e.g.). So you'd be able to encapsulate this pretty much the same way as the Gist shortcode, but instead of inserting GH's JS, you'd insert KaTeX and emit the LaTeX verbatim. Then you also wouldn't have to rely on KaTeX's autoparser, but can invoke it manually for inserted shortcode divs, just like the Gists are doing (the JS loaded from GH Gist is DOM-modifying, too, btw).

Basically emit (pseudocode):

```html
<div id="latex-1">plain latex goes here</div>
<script>render(id("latex-1"))</script>
```

once per section and then include katex JS and CSS once per page.

> what notation is the most common?

This is not relevant if you go the shortcode route. Though if you use KaTeX's autoparser, it inherits the de facto standard notation, which is fenced by one or more `$`. This is used on MathJax, KaTeX and tons of sides like MathOverflow (and StackExchange in general). This is potentially dangerous if people include these symbols in their prose and it gets mistaken as LaTeX. I think the shortcode route is a better way to do that.

**Comment by rhymes** [comment]
Thanks a lot for the explanation and clarification @turbo, I really appreciate it!

**Comment by nestedsoftware** [comment]
I recently created a copy of my dev.to blog hosted on github pages. I am using jekyll to process the markdown files. It turns out there is a jekyll plugin for katex that I am going to try out:

https://github.com/linjer/jekyll-katex

I can update here if it is working okay on my end. If so, hopefully it would not be too hard to port this plugin to dev.to. I've looked a little bit at the source for some of the dev.to liquid preview tags, and it seems to be similar to the way jekyll does plugins.

**Comment by nestedsoftware** [comment]
I tried out jekyll-katex and it works great out of the box! It processes the math notation on the server, so there is no need to parse the math on-the-fly in the browser itself. 

To get this working in jekyll: 

- Install the `jekyll-katex` gem per the readme (https://github.com/linjer/jekyll-katex)
- Download the `katex` source from `https://github.com/KaTeX/KaTeX/releases` (I got version 0.11.1 which is the current default in the plugin).
- Unpack the zip or tarball and go to the `KaTeX-0.11.1/KaTeX-0.11.1/dist` directory. Copy `katex.min.css` and the `fonts` folder and paste them both to the jekyll project under `assets/css`.
- Add the katex css to the jekyll project, e.g. in the `default.html` file, i.e. `<link rel="stylesheet" href="/assets/css/katex.min.css">`
- Add a liquid tag to the markdown for an article, e.g. `Let's have some math! {% katex %} c = \pm\sqrt{a^2 + b^2} {% endkatex %}`
- Run `bundle exec jekyll build` and `bundle exec jekyll serve`
- Go to the generated site and witness the pure mathy joy!

Since dev.to is also ruby, hopefully it would not be too bad to incorporate this logic for dev.to also!

**Comment by a7v8x** [comment]
This would be super helpful, I am thinking about reposting more machine learning stuff on dev.to, but this feature is making my workflow for importing original posts with math very difficult. I agree with implementing it with KaTex, having great experiences with that as well.

**Comment by thphuc** [comment]
Do we have any plan to make it easier for users? I believe Latex/Katex are very complicated to almost users

**Comment by maestromac** [comment]
Hello @thphuc , we currently don't have any plan on that. If you would like to push for this request, you can submit a discussion post on https://github.com/forem/forem/discussions
